Framework/API/src/InfoComponentVisitor.cpp

```diff
@@ -0,0 +1,128 @@
+#include "MantidAPI/InfoComponentVisitor.h"
+#include "MantidAPI/DetectorInfo.h"
+#include "MantidGeometry/IComponent.h"
+#include "MantidGeometry/ICompAssembly.h"
+#include "MantidGeometry/IDetector.h"
+
+#include <numeric>
+#include <algorithm>
+
+namespace Mantid {
+namespace API {
+
+using namespace Mantid::Geometry;
+
+InfoComponentVisitor::InfoComponentVisitor(
+    const size_t nDetectors,
+    std::function<size_t(const Mantid::detid_t)> mapperFunc)
+    : m_componentIds(nDetectors, nullptr),
+      m_detectorIdToIndexMapperFunction(mapperFunc) {
+  m_detectorIndices.reserve(nDetectors);
+}
+
+/**
+ * @brief InfoComponentVisitor::registerComponentAssembly
+ * @param assembly : ICompAssembly being visited
+ */
+void InfoComponentVisitor::registerComponentAssembly(
+    const ICompAssembly &assembly) {
+
+  std::vector<IComponent_const_sptr> assemblyChildren;
+  assembly.getChildren(assemblyChildren, false /*is recursive*/);
+
+  const size_t detectorStart = m_detectorIndices.size();
+  for (const auto &child : assemblyChildren) {
+    // register everything under this assembly
+    child->registerContents(*this);
+  }
+  const size_t detectorStop = m_detectorIndices.size();
+
+  m_ranges.emplace_back(std::make_pair(detectorStart, detectorStop));
+
+  // For any non-detector we extend the m_componetIds from the back
+  m_componentIds.emplace_back(assembly.getComponentID());
+}
+
+/**
+ * @brief InfoComponentVisitor::registerGenericComponent
+ * @param component : IComponent being visited
+ */
+void InfoComponentVisitor::registerGenericComponent(
+    const IComponent &component) {
+  /*
+   * For a generic leaf component we extend the component ids list, but
+   * the detector indexes entries will of course be empty
+   */
+  m_ranges.emplace_back(std::make_pair(0, 0)); // Represents an empty range
+  m_componentIds.emplace_back(component.getComponentID());
+}
+
+/**
+ * @brief InfoComponentVisitor::registerDetector
+ * @param detector : IDetector being visited
+ */
+void InfoComponentVisitor::registerDetector(const IDetector &detector) {
+
+  const auto detectorIndex =
+      m_detectorIdToIndexMapperFunction(detector.getID());
+
+  /* Unfortunately Mantid supports having detectors attached to an
+   * instrument that have an an invalid or duplicate detector id.
+   * We do not register detectors with the same id twice.
+   */
+  if (m_componentIds[detectorIndex] == nullptr) {
```

Comment:
I suppose this breaks if the order in which detectors were added to the instrument is not equal to the order in which components are visited. Did you have to fix any tests due to this?